### DISCUSSION TITLE
How to supply Personal Access Token via the session_auth to the Client and proxy config

### DISCUSSION THREAD

**Discussion by raffaellof** [dennisvang/tufup#153] [Q&A] [answered]
Maybe a stupid question, but I'm struggling trying to supply a Personal Access Token from myapp to a private github repository. Everytime app check_for_updates, it returns a 404 log error.

`tufup.client - WARNING - Cannot refresh metadata: 404 Client Error: Not Found for url: https://api.github.com/repos/.../timestamp.json`

I tried: 

```python
from requests.auth import AuthBase

class HTTPBearerToken(AuthBase):
    def __init__(self, token):
        self.token = token

    def __call__(self, r):
        r.headers['authorization'] = f'Bearer {self.token}'
        return r

client = Client(
            app_name=settings.APP_NAME,
            app_install_dir=settings.INSTALL_DIR,
            current_version=settings.APP_VERSION,
            metadata_dir=settings.METADATA_DIR,
            metadata_base_url=settings.METADATA_BASE_URL,
            target_dir=settings.TARGET_DIR,
            target_base_url=settings.TARGET_BASE_URL,
            refresh_required=False,
            session_auth={'https://api.github.com/': HTTPBearerToken(token)}
        )
```

No success... Of course, making a standard request (with requests.get) works properly (both token and repo are available). Maybe proxy configuration? Why error involves timestamp.json only? (This is a blocking error, I don't know if any other files are actually found)

Is there any way to pass proxy config to the client? Any other suggestions? Thanks

**Comment by dennisvang** [comment]
Hi @raffaellof , sorry for the late answer. 

From the [docs][1]:

>If you try to use a REST API endpoint without a token or with a token that has insufficient permissions, you will receive a `404 Not Found` or `403 Forbidden` response. Authenticating with invalid credentials will initially return a `401 Unauthorized` response.

My first hunch would be an error in the url used for `session_auth`, viz. there should not be a trailing slash. As a result, tufup sets an empty `session.auth` (where `session` is a [`requests.Session`][4] object):

https://github.com/dennisvang/tufup/blob/0283ebb420e7aa3408757dd9d36cd45c455c6490/src/tufup/client.py#L401

Could you try removing the trailing slash from your code, as follows?

```python
session_auth={'https://api.github.com': HTTPBearerToken(token)}
```
Here's the relevant part of the docstring for the `AuthRequestsFetcher`:

https://github.com/dennisvang/tufup/blob/0283ebb420e7aa3408757dd9d36cd45c455c6490/src/tufup/client.py#L346-L361

If that does not help, perhaps it has to do with the permissions for your PAT. Here's a list of required permissions for [fine-grained PATs][2], per API endpoint, and here a list of scopes for [classic PATs][3]. Another possible cause for a 404 could arise with classic PAT and SSO, according to the [docs][5]:

>If you do not authorize your personal access token (classic) for SAML SSO before you try to use it to access a single organization that enforces SAML SSO, you may receive a `404 Not Found`

[1]: https://docs.github.com/en/rest/authentication/authenticating-to-the-rest-api?apiVersion=2022-11-28#failed-login-limit
[2]: https://docs.github.com/en/rest/authentication/permissions-required-for-fine-grained-personal-access-tokens?apiVersion=2022-11-28
[3]: https://docs.github.com/en/apps/oauth-apps/building-oauth-apps/scopes-for-oauth-apps#available-scopes
[4]: https://requests.readthedocs.io/en/latest/api/#request-sessions
[5]: https://docs.github.com/en/rest/authentication/authenticating-to-the-rest-api?apiVersion=2022-11-28#personal-access-tokens-and-saml-sso

**Reply by raffaellof** [reply]
First of all thanks for your answer, I managed to reach the file by adding more headers (as described here https://docs.github.com/en/rest/repos/contents?apiVersion=2022-11-28 ) and using another network configuration, but still having problems behind the proxy, isn't there a way to add proxy to the session_auth as well?

**Reply by dennisvang** [reply]
Good to hear you managed to reach the file.

I assume that means you removed the *trailing slash* from the `session_auth` key, as suggested [above][1]?

I'm fairly certain that's [necessary][2] for authentication to work.

[1]: https://github.com/dennisvang/tufup/discussions/153#discussioncomment-9891346
[2]: https://gist.github.com/dennisvang/e699e1ae28dbe473193f5487f3becec5

**Reply by dennisvang** [reply]
>... by adding more headers ... and using another network configuration ...

It would be helpful to know the details of your workaround.

**Reply by dennisvang** [reply]
>... still having problems behind the proxy, ...

Could you explain which specific problems you are facing? (please be explicit here)

>... isn't there a way to add proxy to the session_auth as well?

Currently, no, not out-of-the-box. I would consider that an advanced feature. 

However, if you really need it, you could subclass `tufup.client.AuthRequestsFetcher` to set e.g. [`session.proxies`][1], and then [monkey-patch][2] `tufup.client.Client.__init__` to use your custom fetcher.

[1]: https://requests.readthedocs.io/en/latest/api/#requests.Session.proxies
[2]: https://docs.python.org/3/library/unittest.mock.html#the-patchers

**Comment by raffaellof** [comment]
Here I am... much appreciated your help

> I assume that means you removed the _trailing slash_ from the `session_auth` key, as suggested [above](https://github.com/dennisvang/tufup/discussions/153#discussioncomment-9891346)?

Sure, I did, it was a typo, my bad

> It would be helpful to know the details of your workaround.

Subclass of requests.auth.AuthBase to add headers like so:

```python
class HttpsGitTokenAuth(AuthBase):
    """Subclass of requests.auth.AuthBase to add custom headers"""
    def __init__(self, token):
        self.token = token

    def __call__(self, r):
        r.headers['authorization'] = f'Bearer {self.token}'
        r.headers['accept'] = 'application/vnd.github.raw+json'
        r.headers['X-GitHub-Api-Version'] = '2022-11-28'
        return r
```
This works for me, when the app runs in a network **without** proxies

> Could you explain which specific problems you are facing? (please be explicit here)

Still facing an error trying downloading files from the repo, when my app runs in a network **behind a proxy**

`tufup.client - WARNING - Cannot refresh metadata: Failed to download https://api.github.com/repos...`

Then I followed your advice (thanks for putting me on the right way)

>...if you really need it, you could subclass tufup.client.AuthRequestsFetcher to set e.g. [session.proxies](https://requests.readthedocs.io/en/latest/api/#requests.Session.proxies), and then [monkey-patch](https://docs.python.org/3/library/unittest.mock.html#the-patchers) tufup.client.Client.__init__ to use your custom fetcher.

Subclass of tufup.client.AuthRequestsFetcher to add my custom session.proxies (overridden methods __init__ and _get_session)

```python
class AuthRequestsFetcherProxyMode(AuthRequestsFetcher):
    """Subclass of tufup.client.AuthRequestsFetcher to add custom proxies configuration"""
    def __init__(
            self,
            session_auth: Optional[Dict[str, Union[Tuple[str, str], AuthBase]]] = None
    ) -> None:
        super().__init__(session_auth=session_auth)

        # My custom proxies
        self.proxies = {  # Same proxy for both protocols
            'http': 'http://proxy:port',
            'https': 'http://proxy:port'
        }

    def _get_session(self, url: str) -> requests.Session:
        # set the Session.proxies attribute for the specified server, if available
        session = super()._get_session(url=url)
        session.proxies = self.proxies
        return session
```

...and monkey-patch the tufup.client.Client.__init__ in the simplest way I figured out (context-manager):
```python

        with mock.patch(target='tufup.client.AuthRequestsFetcher', new=AuthRequestsFetcherProxyMode):
            client = Client(
                app_name=settings.APP_NAME,
                app_install_dir=settings.INSTALL_DIR,
                current_version=settings.APP_VERSION,
                metadata_dir=settings.METADATA_DIR,
                metadata_base_url=settings.METADATA_BASE_URL,
                target_dir=settings.TARGET_DIR,
                target_base_url=settings.TARGET_BASE_URL,
                refresh_required=False,
                session_auth={'https://api.github.com': HttpsGitTokenAuth(git_token)}
            )
```
...client object initialized successfully!

Unfortunately I didn't solve the problem... and i don't know why! When my app runs behind a proxy and checks for updates, keep logging **Failed Download** error. I still think something goes wrong in the proxy configuration step, still looking for a fix :(

**Comment by dennisvang** [comment]
>... Unfortunately I didn't solve the problem... and i don't know why! When my app runs behind a proxy and checks for updates, keep logging **Failed Download** error. I still think something goes wrong in the proxy configuration step, still looking for a fix :(

@raffaellof Sorry to hear that.

Could you post the relevant part of the log, showing the actual "Failed Download" error?

**Comment by dennisvang** [comment, marked as answer]
@raffaellof On a side note, I just found out there's an easier way to configure proxies for `requests`: 

You can simply set the [environment variables][1] `HTTP_PROXY`, `HTTPS_PROXY`, etc.

At least this implies you should not need to subclass the `AuthRequestsFetcher`.

It may also help with the download issue.

[1]: https://requests.readthedocs.io/en/latest/user/advanced/#proxies

**Comment by raffaellof** [comment]
I took some time to do some tests... the only way to make my app working properly, even behind a proxy, is by setting environment variables as suggested [above](https://github.com/dennisvang/tufup/discussions/153#discussioncomment-9922727). 
Environment variables which (obviously) can be set dynamically within the app itself, so I got rid of the `AuthRequestsFetcher` subclass (and also monkeypatch). 

```python
os.environ['HTTP_PROXY'] = f'http://{user}:{password}@{PROXY}:{PORT}'
os.environ['HTTPS_PROXY'] = f'http://{user}:{password}@{PROXY}:{PORT}'
```
P.S. If (as in my case) the proxy requires authorization, credentials should be percent-encoded to avoid conflicts, to do this I used `urllib`

```python
import urllib.parse

user = urllib.parse.quote(user)
password = urllib.parse.quote(password)
```

Problem solved. Thanks for your precious help @dennisvang

**Reply by dennisvang** [reply]
@raffaellof happy to hear the problem is solved. Thanks for the useful feedback!
